core/src/main/java/org/apache/iceberg/BaseTransaction.java

```diff
@@ -254,16 +256,13 @@ private void commitCreateTransaction() {
               ((SnapshotProducer) update).cleanAll();
             }
           });
-
-      throw e;
-
-    } finally {
       // create table never needs to retry because the table has no previous state. because retries are not a
       // concern, it is safe to delete all of the deleted files from individual operations
       Tasks.foreach(deletedFiles)
           .suppressFailureWhenFinished()
           .onFailure((file, exc) -> LOG.warn("Failed to delete uncommitted file: {}", file, exc))
           .run(ops.io()::deleteFile);
+      throw e;
```

Comment:
The changes here aren't correct. Before, the deleted files were removed after the commit in `finally`, regardless of the outcome. I understand that this change is intended to avoid running the block when `CommitStateUnknownException` is thrown. But this also doesn't run the delete when no exception is thrown, which is a bug that will leak files.

I also don't think that it is correct to not run the deletes when `CommitStateUnknownException` is thrown. That indicates that the commit may have succeeded and the new files written by the commit should not be cleaned up. But this list of deleted files is the set of files that were created but are no longer used even if the commit succeeds. This is primarily files from previous commit attempts. They should still be deleted even if the commit state is unknown.

I think the only change that is actually needed is the addition of the catch block for `CommitStateUnknownException`.

Reply:
Interesting, @jfz Can you additional add checks in the tests to make sure these files which are supposed to be deleted are actually removed? My guess is we don't actually have any tests for this behavior since everything is currently passing.

Reply:
@rdblue Did some debug and I think you are right, and for the case of `CommitStateUnknownException` we are not adding files generate by the very last commit attempt to `deletedFiles` so they will not be deleted. Pushed a fix.

Reply:
Thanks, this looks correct to me now.

It would be great to look into the tests like @RussellSpitzer suggests, but I don't think that needs to block this PR. Russell, what do you think?

Reply:
I added basic cleanup checks by counting number of files left.

Reply:
Confirmed, thanks for adding those checks!